File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.cs
using System;
using System.Collections.Generic;

public class Solution {
    private static bool[,] visited;
    private static int[] dx = {1, -1, 0, 0};
    private static int[] dy = {0, 0, 1, -1};
    public int[] solution(string[] maps) {
        List<int> list = new List<int>();
        int row = maps.Length;
        int col = maps[0].Length;
        visited = new bool[row, col];
        
        for(int i = 0; i < row; i++)
        {
            for(int j = 0; j < col; j++)
            {
                if(!visited[i, j] && maps[i][j] != 'X')
                {
                    int result = DFS(i, j, row, col, maps);
                    list.Add(result);
                }
            }
        }
        
        if(list.Count == 0) return new int[] {-1};
        list.Sort();
        return list.ToArray();
    }
    
    private static int DFS(int x, int y, int row, int col, string[] maps)
    {
        var stack = new Stack<(int, int)>();
        stack.Push((x, y));
        visited[x, y] = true;
        int size = maps[x][y] - '0';
        
        while(stack.Count > 0)
        {
            var (cx, cy) = stack.Pop();
            
            for(int i = 0; i < 4; i++)
            {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                if(nx >= 0 && nx < row && ny >= 0 && ny < col &&
                  !visited[nx, ny] && maps[nx][ny] != 'X')
                {
                    visited[nx, ny] = true;
                    size += maps[nx][ny] - '0';
                    stack.Push((nx, ny));
                }
            }
        }
        return size;
    }
}

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int dy[] = {-1, 0, 1, 0};
int dx[] = {0, 1, 0, -1};

bool visited[104][104];
vector<int> answer;

int dfs(int y, int x, vector<string> &maps)
{
    visited[y][x] = true;
    
    int sum = maps[y][x] - '0';
    
    for(int i = 0; i < 4; i++)
    {
        int ny = y + dy[i];
        int nx = x + dx[i];
        
        if(ny < 0 || nx < 0 || ny >= maps.size() || nx >= maps[0].size())
            continue;
        if(visited[ny][nx]) continue;
        if(maps[ny][nx] == 'X') continue;
        
        sum += dfs(ny, nx, maps);
    }
    return sum;
}
vector<int> solution(vector<string> maps) {
    
    for(int i = 0; i < maps.size(); i++)
    {
        for(int j = 0; j < maps[0].size(); j++)
        {
            if(isdigit(maps[i][j]) && !visited[i][j])
            {
                answer.push_back(dfs(i, j, maps));
            }
        }
    }
    if(answer.empty()) return {-1};
    sort(answer.begin(), answer.end());
    return answer;
}